File: 2020/advent-of-code/src/advent_of_code/day21.py
#!/usr/bin/env python3

import re

def parse_line(line):
    matches = re.match(r"(.+) \(contains (.+)\)", line)
    return {"ingredients": set(matches[1].split()), "allergens": set(matches[2].split(", "))}

if __name__ == "__main__":

    foods = [parse_line(line) for line in open("../../resources/advent_of_code/2020/day21/input")]
    allergens = {}

    for _ in range(len(foods)):
        for a in set.union(*[food['allergens'] for food in foods]):
            ingredients = set.intersection(*[f['ingredients'] for f in foods if a in f['allergens']])
            if len(ingredients) == 1:
                i = list(ingredients)[0]
                allergens[a] = i
                for f in foods:
                    if a in f['allergens']:
                        f['allergens'].remove(a)
                    if i in f['ingredients']:
                        f['ingredients'].remove(i)

    answer1 = sum([len(f['ingredients']) for f in foods])
    print("ANSWER1:", answer1)

    answer2 = ",".join([allergens[i] for i in sorted(allergens.keys())])
    print("ANSWER2:", answer2)

File: 2020/advent-of-code/src/advent_of_code/day21.clj
(ns advent-of-code.day21
  (:require [advent-of-code.util :as util]))

(defn parse-line [line]
  (if-let [[_ ingredients allergens] (re-matches #"(.+) \(contains (.+)\)" line)]
    (let [ingredients (clojure.string/split ingredients #" ") allergens (clojure.string/split allergens #", ")]
      {:ingredients (set ingredients) :allergens (set allergens)})))

(defn parse-input []
  nil)

(defn parse-test-input []
  (util/load-test-input 2020 21 parse-line))


(defn eliminate-ingredients [foods]
  (let [allergens (apply clojure.set/union (map :allergens foods))]
    fuuu))

(eliminate-ingredients (parse-test-input))

